Add support for litecoin p2sh-segwit and bech32 address types

bisoge opened the pull request.

cdecker commented on bitcoin/base58.c:
```diff
@@ -92,9 +92,10 @@ bool p2sh_from_base58(bool *test_net,
 		return false;
 
-	if (version == 196)
+	/* 196 BTC testnet, 58 LTC testnet, 5 BTC mainnet, 50 LTC mainnet */
```
These values could also be moved into chainparams, and here we'd just iterate over all known chainparams and set `*test_net = cp->testnet` if `cp->p2sh_version` matches.

bisoge replied:
Agreed, done in next commit.

bisoge commented:
Rebased and squashed into single commit due to conflicts with #1913. Also added legacy address support and fixed invoice and p2sh-segwit address generation.
